### PR TITLE
Add Crescent APR calculation to api-server

### PR DESCRIPTION
**Pull request by 0xbala-k**
Crescent implements its own mint module and the apr is calculated as follows
apr=(1-budget rate)*(1-tax)*CurrentInflationAmount/Bonded tokens

### CODE REVIEW

File: api/chains/chains.go [hunk 1]
```diff
@@ -1234,6 +1241,211 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
...
+			chain.ChainName,
+		)
+		_ = c.Error(e)
+
+		return "", err
+	}
+
+	var budgetRate sdktypes.Dec
+	for _, budget := range budgetParamsData.Params.Budgets {
+		if budget.Name == "budget-ecosystem-incentive" || budget.Name == "budget-dev-team" {
```
Comment (sgerogia) [inline api/chains/chains.go]: Extract as constants?

File: api/chains/chains.go [hunk 2]
```diff
@@ -1234,6 +1241,211 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
 	}
 }
 
+// apr=(1-budget rate)*(1-tax)*CurrentInflationAmount/Bonded tokens
+func getCrescentAPR(c *gin.Context, chain cns.Chain, bondedTokens sdktypes.Dec, client sdkutilities.Client) (string, error) {
```
Comment (sgerogia) [inline api/chains/chains.go:1248]: Can the method be broken up into data fetching and calculation?
This way you can mock fetched data add a unit test on the calculation logic (e.g. edge cases, invalid data)?

**Review by sgerogia — CHANGES_REQUESTED** [review]

File: api/chains/chains.go [hunk 3]
```diff
@@ -1234,6 +1241,211 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
...
+			).WithLogContext(
+				fmt.Errorf("cannot convert end time to time: %w", err),
+				"name",
+				chain.ChainName,
+			)
+			_ = c.Error(e)
+
+			return "", err
+		}
+		if StartTime.After(now) && EndTime.Before(now) {
```
Comment (tbruyelle) [inline api/chains/chains.go:1493]: If this condition is never true, you will have a panic here :

https://github.com/EmerisHQ/demeris-api-server/blob/48ec73bd10f78577cf9ae1cf8c48b4130f2a4c69/api/chains/chains.go#L1445

So the question is, are you sure you have at least one `schedule` that meets this requirement?

Comment (0xbala-k) [reply api/chains/chains.go:1493]: initialized `currentInflationAmount` to 0. So, if no schedule is found the APR will be zero.

**Review by sgerogia — APPROVED** [review]

File: api/chains/chains.go [hunk 4]
```diff
@@ -1234,6 +1244,274 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
...
+
+	OneDec := sdktypes.NewDec(1)
+	apr := OneDec.Sub(tax).Mul(OneDec.Sub(budgetRate)).Mul(currentInflationAmount).Quo(bondedTokens)
+	return apr.String(), nil
+}
+
+func getBudgetRate(c *gin.Context, chain cns.Chain, client sdkutilities.Client) (sdktypes.Dec, error) {
+	var budgetRate sdktypes.Dec
+
+	budgetParamsResp, err := client.BudgetParams(context.Background(), &sdkutilities.BudgetParamsPayload{
```
Comment (pitasi) [inline api/chains/chains.go]:
```suggestion
	budgetParamsResp, err := client.BudgetParams(c.Request.Context(), &sdkutilities.BudgetParamsPayload{
```

File: api/chains/chains.go [hunk 5]
```diff
@@ -1234,6 +1244,274 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
...
+		}
+	}
+
+	return budgetRate, nil
+}
+
+func getTax(c *gin.Context, chain cns.Chain, client sdkutilities.Client) (sdktypes.Dec, error) {
+	var tax sdktypes.Dec
+
+	distributionParamsResp, err := client.DistributionParams(context.Background(), &sdkutilities.DistributionParamsPayload{
```
Comment (pitasi) [inline api/chains/chains.go]:
```suggestion
	distributionParamsResp, err := client.DistributionParams(c.Request.Context(), &sdkutilities.DistributionParamsPayload{
```

File: api/chains/chains.go (hunk 3 above)
Comment (tbruyelle) [reply api/chains/chains.go:1493]: perfect thx

File: api/chains/chains.go [hunk 6]
```diff
@@ -1234,6 +1244,274 @@ func getAPR(c *gin.Context, sdkServiceClients sdkservice.SDKServiceClients) stri
...
+
+		return tax, err
+	}
+	return tax, nil
+}
+
+func getCurrentInflationAmount(c *gin.Context, chain cns.Chain, client sdkutilities.Client) (sdktypes.Dec, error) {
+	currentInflationAmount := sdktypes.NewDec(0)
+
+	mintParamsResp, err := client.MintParams(context.Background(), &sdkutilities.MintParamsPayload{
```
Comment (pitasi) [inline api/chains/chains.go]:
```suggestion
	mintParamsResp, err := client.MintParams(context.Background(), &sdkutilities.MintParamsPayload{
```
```suggestion
	mintParamsResp, err := client.MintParams(c.Request.Context(), &sdkutilities.MintParamsPayload{
```

**Review by pitasi — CHANGES_REQUESTED** [review]

**Review by pitasi — APPROVED** [review]
lgtm
